corehq/apps/userreports/tests/test_json_query_expressions.py

```diff
@@ -0,0 +1,455 @@
+        ExpressionFactory.from_spec(expression)(items_ex, map_ex)
+
+
+@generate_cases([
+    (
+        {'items': [{'key': 'v1'}, {'key': 'v2'}]},
+        {'type': 'property_name', 'property_name': 'items'},
+        {'type': 'property_name', 'property_name': 'key'},
+        ['v1', 'v2']
+    ),
+    (
+        [{'key': 'v1'}, {'key': 'v2'}],
+        {'type': 'identity'},
+        {'type': 'property_name', 'property_name': 'key'},
+        ['v1', 'v2']
+    ),
+    (
+        {'items': [{'key': 'v1'}, {'key': 'v2'}]},
+        {'type': 'property_name', 'property_name': 'items'},
+        {'type': 'identity'},
+        [{'key': 'v1'}, {'key': 'v2'}],
+    ),
+    # literal
+    (
+        [],
+        [{'key': 'v1'}, {'key': 'v2'}],
+        {'type': 'identity'},
+        [{'key': 'v1'}, {'key': 'v2'}],
+    ),
+    # map_expression can return empty on an item
+    (
+        {'items': [{'key': 'v1'}, {'no_key': 'v2'}]},
+        {'type': 'property_name', 'property_name': 'items'},
+        {'type': 'property_name', 'property_name': 'key'},
+        ['v1', None]
+    ),
+    # if items_expression returns empty, an empty list is returned
+    (
+        {'items': [{'key': 'v1'}, {'no_key': 'v2'}]},
+        {'type': 'property_name', 'property_name': 'no_match'},
+        {'type': 'property_name', 'property_name': 'key'},
+        []
+    ),
+])
+def test_map_items_basic(self, doc, items_ex, map_ex, expected):
+    expression = ExpressionFactory.from_spec({
+        'type': 'map_items',
+        'items_expression': items_ex,
+        'map_expression': map_ex
+    })
+    self.assertEqual(expression(doc), expected)
+
+
+@generate_cases([
+    ([{'key': 'v1'}, {'key': 'v2'}], None),
+    ({'type': 'identity'}, {}),
+    ({'type': 'identity'}, 'invalid_agg_fn'),
+])
+def test_reduce_items_bad_spec(self, items_ex, reduce_ex):
+    expression = {
+        'type': 'reduce_items',
+    }
+    if items_ex is not None:
+        expression['items_expression'] = items_ex
+    if reduce_ex is not None:
+        expression['aggregation_fn'] = reduce_ex
+
+    with self.assertRaises(BadSpecError):
+        ExpressionFactory.from_spec(expression)(items_ex, reduce_ex)
+
+
+@generate_cases([
+    (
+        ['a', 'b'],
+        {'type': 'identity'},
+        'count',
+        2,
+    ),
+    (
+        {'items': ['a', 'b']},
+        {'type': 'property_name', 'property_name': 'items'},
+        'count',
+        2,
+    ),
+    (
+        {'items': ['a', 'b']},
+        {'type': 'property_name', 'property_name': 'no_match'},
+        'count',
+        0,
+    ),
+    (
+        [],
+        {'type': 'identity'},
+        'count',
+        0,
+    ),
+    (
+        [1, 2],
+        {'type': 'identity'},
+        'sum',
+        3
+    ),
+    (
+        [1, 2],
+        {'type': 'identity'},
+        'first_item',
+        1
+    ),
+    (
+        [1, 2],
+        {'type': 'identity'},
+        'last_item',
+        2
+    ),
+])
```

Comment:
should also add tests for some of the error conditions above. including the ones where the items expression doesn't return an iterable as well as things like `first_item` on an empty list